# Chapter 3, Data Cleaning

In [88]:
import pickle
import pandas as pd
import numpy as np
from dotenv import load_dotenv, find_dotenv
import os

In [89]:
def check_for_missing_vals(df):
    for x in list(set(df.columns.get_level_values(0))):
        if sum(list(df[x].isna().sum())) != 0:
            print(x)
            print(list(df[x].isna().sum()))

In [90]:
load_dotenv(find_dotenv())
data_base_dir = os.environ.get('DATA_DIR_BASE_PATH')
fname = data_base_dir + '/processed/djia_raw_2019-12-11.pkl'
djia_raw = pickle.load(open(fname, 'rb'))

In [92]:
djia_raw['MCD'].tail()

Open        High         Low       Close  \
Epoch                                                                       
2019-11-27 00:00:00+00:00  194.259995  196.669998  193.919998  196.300003   
2019-11-29 00:00:00+00:00  196.509995  196.539993  194.270004  194.479996   
2019-12-02 00:00:00+00:00  195.380005  195.589996  193.779999  195.179993   
2019-12-03 00:00:00+00:00  194.300003  195.039993  192.350006  193.119995   
2019-12-04 00:00:00+00:00  193.589996  194.460007  192.889999  194.309998   

                            Volume  
Epoch                               
2019-11-27 00:00:00+00:00  4136783  
2019-11-29 00:00:00+00:00  2815480  
2019-12-02 00:00:00+00:00  3198902  
2019-12-03 00:00:00+00:00  4078751  
2019-12-04 00:00:00+00:00  2217253

In [93]:
# Look for missing values in the OHLCV data of every symbol and return symbol if missing values are found
check_for_missing_vals(djia_raw)

DOW
[1087, 1087, 1087, 1087, 1087]


Looks like there are 1087 missing values for Dow Chemicals

In [94]:
djia_raw['DOW'].iloc[1086:, :]

Open       High        Low      Close  \
Epoch                                                                   
2019-04-01 00:00:00+00:00        NaN        NaN        NaN        NaN   
2019-04-02 00:00:00+00:00  55.119999  57.389999  54.299999  56.250000   
2019-04-03 00:00:00+00:00  57.080002  57.200001  55.509998  56.880001   
2019-04-04 00:00:00+00:00  57.200001  60.520000  57.099998  59.709999   
2019-04-05 00:00:00+00:00  58.389999  58.980000  56.270000  57.240002   
...                              ...        ...        ...        ...   
2019-11-27 00:00:00+00:00  54.680000  54.900002  53.900002  54.279999   
2019-11-29 00:00:00+00:00  54.009998  54.279999  53.119999  53.369999   
2019-12-02 00:00:00+00:00  53.000000  53.750000  52.889999  53.160000   
2019-12-03 00:00:00+00:00  51.959999  52.009998  51.049999  51.820000   
2019-12-04 00:00:00+00:00  53.490002  53.490002  52.209999  52.279999   

                               Volume  
Epoch                                  
2019-04-01 00:00:00+00:00         NaN  
2019-04-02 00:00:00+00:00  17414153.0  
2019-04-03 00:00:00+00:00  15066169.0  
2019-04-04 00:00:00+00:00  13311868.0  
2019-04-05 00:00:00+00:00  17000788.0  
...                               ...  
2019-11-27 00:00:00+00:00   3926099.0  
2019-11-29 00:00:00+00:00   1887577.0  
2019-12-02 00:00:00+00:00   3364773.0  
2019-12-03 00:00:00+00:00   3990311.0  
2019-12-04 00:00:00+00:00   3791442.0  

[173 rows x 5 columns]

No data exists for DOW earlier than 2019-04-02. DOW went public on 2019-03-21. This explains the missing values. Fill them with zeros.

In [70]:
djia_raw.fillna(0, inplace=True)

In [71]:
check_for_missing_vals(djia_raw)

No more missing values.

In [95]:
fname = data_base_dir + '/processed/djia_cleaned_2019-12-11.pkl'
pickle.dump(djia_raw, open(fname, 'wb'))